pkg/reconciler/managed/metrics.go

```diff
@@ -0,0 +1,204 @@
+/*
+Copyright 2023 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package managed
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/tools/cache"
+	kmetrics "k8s.io/component-base/metrics"
+	"sigs.k8s.io/controller-runtime/pkg/cluster"
+	"sigs.k8s.io/controller-runtime/pkg/metrics"
+
+	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+	"github.com/crossplane/crossplane-runtime/pkg/errors"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+)
+
+func init() { //nolint:gochecknoinits // metrics should be registered once
+	metrics.Registry.MustRegister(drift, mr, mrReady, mrSynced, mrDetected, mrReadyDuration, mrDeletionDuration)
+}
+
+const subSystem = "crossplane"
+
+var (
+	drift = prometheus.NewHistogramVec(prometheus.HistogramOpts{ //nolint:gochecknoglobals // metrics should be registered once in init
+		Subsystem: subSystem,
+		Name:      "resource_drift_seconds",
+		Help:      "ALPHA: How long since the previous successful reconcile when a resource was found to be out of sync; excludes restart of the provider",
+		Buckets:   kmetrics.ExponentialBuckets(10e-9, 10, 10),
+	}, []string{"group", "kind"})
+
+	mr = prometheus.NewGaugeVec(prometheus.GaugeOpts{ //nolint:gochecknoglobals // metrics should be registered once in init
+		Subsystem: subSystem,
+		Name:      "managed_resource_created",
+		Help:      "The number of managed resources created",
+	}, []string{"gvk", "name", "claim", "composite"})
+
+	mrReady = prometheus.NewGaugeVec(prometheus.GaugeOpts{ //nolint:gochecknoglobals // metrics should be registered once in init
+		Subsystem: subSystem,
+		Name:      "managed_resource_ready",
+		Help:      "The number of managed resources in Ready=True state",
+	}, []string{"gvk", "name", "claim", "composite"})
```

Comment:
I think if I don't have MR name as label, I need to store which MRs that I recorded in a map and check whether I sent a metric for that MR or not. Since these metrics are recorded in a reconciliation loop, it's easy to send duplicate metrics for the same MR and the solution I found is using MR name as label to have unique identifier for metrics. 

Regarding the cardinality, we may remove `claim` and `composite` labels. In the MR metrics [issue](https://github.com/crossplane/crossplane-runtime/issues/555), it's also requested to classify metrics per compositions which requires a change in Crossplane.

Reply:
Unfortunately I'm pretty sure including the MR name in metrics is a non-starter cardinality wise. We would be producing a timeseries for every MR that ever existed during the provider's lifetime (i.e. between restarts). If we think of a control plane with 1,000 MRs, that means:

* Every metric with an MR name label gets 1,000 entries at the `/metrics` http endpoint
* Prometheus (or Prom-compatible scraping agent) must store 1,000 entries in its time series database
* Queries, aggregation rules, etc must process 1,000 entries

I don't think removing claim/composite labels will help much.

Reply:
See also https://prometheus.io/docs/practices/instrumentation/#do-not-overuse-labels

Reply:
+1 for @negz 's point. We can have GVKs in labels, we CANNOT have names. In general, cardinality must never depend on values the user providers (in Crossplane case that is the claim owner).

Another example: Kube had a CVE long time ago with the http user agent in a label. That could be used to DoS kube-apiserver from outside, even non-authenticated.

Reply:
Would we still get a claim level "duration to readiness", and a claim level "duration to resources deleted" metric to allow platform creators to develop SLOs?

Technically that would only need to be 1 time series per claim.

Reply:
@humoflife It would be surprising/unusual to me to have SLOs with that level of granularity.

I would expect the type of SLO you mention to be at the _claim type_ granularity, not individual claim granularity. [^1] This would allow you to graph and alert on situations like "the P95 time for `kind: Cluster` claims to become ready is too high", but not as specific as "the `kind: Cluster` claim named `some-claim-42` took too long to become ready".

Think of monitoring an HTTP server. You'd want timeseries metrics that indicate the aggregate health of the system - the request count, error count, and average duration of requests. You wouldn't emit a timeseries for each HTTP request. You probably want to alert on "too many requests are failing", not "a request failed".

[^1]: I'd also expect such metrics to be emitted by core Crossplane, rather than making providers claim-aware.

Reply:
> This would allow you to graph and alert on situations like "the P95 time for kind: Cluster claims to become ready is too high"

This would need a `crossplane_claim_first_time_to_readiness_seconds{gvk}` timeseries to be emitted by core Crossplane. I would support adding such a timeseries. I'm not sure if it's currently planned.

Reply:
Do you think we should have two different types of metrics for claims and composites or having something like below would work? 
```
crossplane_composite_resource_first_time_to_readiness_seconds{"gvk", "claim-namespace"}
```

Reply:
@ezgidemirel That's a good point, I hadn't considered that you could just decorate the XR metrics with details about their claims (if any).

That said, I do think there's some value in having distinct timeseries for claims. An XR being ready doesn't necessarily mean the associated claim is ready. It's possible that the XR is ready but some issue is preventing the claim from being ready.

Without a distinct timeseries for claims, you'd never know if `Foo` claims were taking a lot longer on average than their corresponding `XFoo` XRs to become ready.